controllers/topology/desired_state.go

```diff
@@ -301,6 +371,93 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, machineDeployme
 	return desiredMachineDeployment, nil
 }
 
+// computeMachineDeploymentVersion calculates the version of the desired machine deployment.
+// The version if calculated using the the state of the current machine deployments, the current control plane
+// and the version defined in the topology.
+func computeMachineDeploymentVersion(s *scope.Scope, desiredControlPlaneState *scope.ControlPlaneState, currentMDState *scope.MachineDeploymentState) (string, error) {
+	desiredVersion := s.Blueprint.Topology.Version
+	// If creating a new machine deployment, we can pick up the desired version
+	// Note: We are not blocking the creation of new machine deployments when
+	// the control plane or any of the machine deployments are upgrading/scaling.
+	if currentMDState == nil || currentMDState.Object == nil {
+		return desiredVersion, nil
+	}
+
+	// Get the current version of the machine deployment.
+	currentVersion := *currentMDState.Object.Spec.Template.Spec.Version
+
+	// Return early if we are not allowed to upgrade the machine deployment.
+	if !s.UpgradeTracker.MachineDeployments.AllowUpgrade() {
+		return currentVersion, nil
+	}
+
+	// Return early if the currentVersion is already equal to the desiredVersion
+	// no further checks required.
+	if currentVersion == desiredVersion {
+		return currentVersion, nil
+	}
+
+	// If the control plane is being created (current control plane is nil), do not perform any machine deployment upgrade in this case.
```

Comment:
nit: shorten these comments (just wrap the lines no need for wording changes) if possible